thread.c

```diff
@@ -545,9 +545,101 @@ static void thread_libevent_process(int fd, short which, void *arg) {
+ */
+static LIBEVENT_THREAD *select_thread_using_napi_id(int sfd)
+{
+    LIBEVENT_THREAD *thread = NULL;
+    int start_tid, i;
+    socklen_t len;
+    int napi_id;
+
+    /* first check if there is pre-existing NAPI_ID => Thread
+     * association, if so - use that thread, otherwise select unassigned
+     * or unassociated thread 
+     */
+#ifndef SO_INCOMING_NAPI_ID
+#define SO_INCOMING_NAPI_ID 56
+#endif /* SO_INCOMING_NAPI_ID */
+    len = sizeof(napi_id);
+    if (getsockopt(sfd, SOL_SOCKET, SO_INCOMING_NAPI_ID,
+		   &napi_id, &len) == -1) {
+        perror("getsockopt SO_INCOMING_NAPI_ID");
+        /* Disable napi_id based thread selection since socket option returned
+         * error, may be running on older linux kernel
+         */
+        settings.napi_id_based_thread_selection = false;
+        return select_thread();
+    }
+    /* getsockopt didn't fail but napi_id is still unassigned, This is expected if
+     * connection are setup on localhost (means not going thru' underlying
+     * device, hence for those local connection, napi_id is not expected
+     * Non NAPI-IDs are aggregated down to zero by getsockopt call
+     * for socket option SO_INCOMING_NAPI_ID
+     */
+	if (!napi_id) {
+        return select_thread();
+	}
+
+    /* find thread which has pre-existing association with napi_id,
+     * if there is pre-existing association of thread => napi_id,
+     * use that thread
+     */
+    for (int i = 0; i < settings.num_threads; i++) {
+        thread = threads + i;
+        if (thread->napi_id == napi_id) {
+            /* found associated thread based on NAPI_ID */
+            return thread;
+        }
+    }
+    /* if thread with pre-existing association of thread => napi_id is not
+     * found, select next available thread based that is not associated
+     * with any NAPI-ID 
+     */
+    start_tid = (last_thread + 1) % settings.num_threads;
+		
+    /* find unclaimed thread based on napi_id */
+    for (i = start_tid; i < settings.num_threads; i++) {
+        thread = threads + i;
+        if (!thread->napi_id) {
+            last_thread = i; /* record index of last selected thread */			
+            /* record napi_id on selected thread to establish
+             * thread => napi_id association. Any connection with same
+             * NAPI_ID is associated with same thread
+             */
+            thread->napi_id = napi_id;
+            return thread;
+        }
+    }
+
+    /* If here means likely that number of unique NAPI_ID are
+     * more than the number of threads and that is fine.
+     * From this point onwards, disable "napi_id_based_thread_selection"
+     * and fallback to default thread selection mechanism
+     * This is needed so that at least application can still continue to work
+     * and apply it's own load-balancing technique
+     */
+    settings.napi_id_based_thread_selection = false;
+    return select_thread();
```

Comment:
I feel like this will give users a mystery. The stat should be added to the stats_settings output, so users can tell if it's actually on or not. Further, we need to document or somehow log/warn/error since it looks like unless you set the worker threads to exactly match the napi_ids it'll just shut itself off immediately.

Reply:
Sure, let me see about stats and find ways to log/warn/errors about threads and napi_ids match

Reply:
Sure, let me look into that.